src/common/os/guid.h

```diff
@@ -46,41 +46,184 @@ struct UUID	// Compatible with Win32 UUID struct layout
+		m_initialized = true;
+	}
+
+	Guid& operator=(const Guid& other)
+	{
+		m_data = other.m_data;
+		m_initialized = other.m_initialized;
+		return *this;
+	}
+
+	bool operator==(const Guid& other) const
+	{
+		return (m_initialized == other.m_initialized &&
+			(!m_initialized || !memcmp(&m_data, &other.m_data, sizeof(UUID))));
+	}
+
+	bool operator!=(const Guid& other) const
+	{
+		return !(*this == other);
+	}
+
+	const UCHAR* getData() const
+	{
+		return reinterpret_cast<const UCHAR*>(&m_data);
+	}
+
+	void assign(const UCHAR* buffer)
+	{
+		memcpy(&m_data, buffer, sizeof(UUID));
+		m_initialized = true;
+	}
 
-	return (result == GUID_FORMAT_ARGS);
-}
+	void clear()
+	{
+		m_initialized = false;
+	}
+
+	bool hasData() const
+	{
+		return m_initialized;
+	}
+
+	bool isEmpty() const
+	{
+		return !m_initialized;
+	}
+
+	void toString(char* buffer) const
+	{
+		sprintf(buffer, GUID_FORMAT,
+			m_data.Data1, m_data.Data2, m_data.Data3,
+			m_data.Data4[0], m_data.Data4[1], m_data.Data4[2], m_data.Data4[3],
+			m_data.Data4[4], m_data.Data4[5], m_data.Data4[6], m_data.Data4[7]);
+	}
+
+	Firebird::string toString() const
+	{
+		Firebird::string result;
+		toString(result.getBuffer(GUID_BUFF_SIZE - 1));
+		return result;
+	}
+
+	Firebird::PathName toPathName() const
+	{
+		Firebird::PathName result;
+		toString(result.getBuffer(GUID_BUFF_SIZE - 1));
+		return result;
+	}
+
+	bool fromString(const char* buffer)
+	{
+		fb_assert(buffer);
+
+		const auto result = sscanf(buffer, GUID_FORMAT,
+			&m_data.Data1, &m_data.Data2, &m_data.Data3,
+			&m_data.Data4[0], &m_data.Data4[1], &m_data.Data4[2], &m_data.Data4[3],
+			&m_data.Data4[4], &m_data.Data4[5], &m_data.Data4[6], &m_data.Data4[7]);
+
+		if (result == GUID_FORMAT_ARGS)
+		{
+			m_initialized = true;
+			return true;
+		}
+
+		return false;
+	}
+
+	bool fromString(const Firebird::string& str)
+	{
+		return fromString(str.nullStr());
+	}
+
+	void copyTo(UUID& ptr) const
+	{
+		ptr = m_data; // copy struct by value
+	}
+
+	void copyTo(UCHAR* ptr) const
+	{
+		memcpy(ptr, &m_data, sizeof(UUID));
+	}
+
+	// Convert platform-dependent GUID into platform-independent form according to RFC 4122
+
+	void convert(UCHAR* data) const
+	{
+		data[0] = (m_data.Data1 >> 24) & 0xFF;
+		data[1] = (m_data.Data1 >> 16) & 0xFF;
+		data[2] = (m_data.Data1 >> 8) & 0xFF;
+		data[3] = m_data.Data1 & 0xFF;
+		data[4] = (m_data.Data2 >> 8) & 0xFF;
+		data[5] = m_data.Data2 & 0xFF;
+		data[6] = (m_data.Data3 >> 8) & 0xFF;
+		data[7] = m_data.Data3 & 0xFF;
+		data[8] = m_data.Data4[0];
+		data[9] = m_data.Data4[1];
+		data[10] = m_data.Data4[2];
+		data[11] = m_data.Data4[3];
+		data[12] = m_data.Data4[4];
+		data[13] = m_data.Data4[5];
+		data[14] = m_data.Data4[6];
+		data[15] = m_data.Data4[7];
+	}
+
+	// Generate new GUID
+
+	static Guid generate()
+	{
+		Guid guid;
+		GenerateGuid(&guid.m_data);
+		guid.m_initialized = true;
+		return guid;
+	}
+
+private:
+	UUID m_data;
+	bool m_initialized = false;
```

Comment:
Why have `initialized` state instead of wrap it in `std::optional` when this functionality is required?

Reply:
Sorry, the original code comes from times when we didn't allow `std::optional` ;-) Will rework, just need to figure out what would be better - use `optional` as data or derive `Guid` from `optional`.

Reply:
Reworked.